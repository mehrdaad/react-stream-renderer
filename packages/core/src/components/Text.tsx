import React from 'react';
import Paragraph from '../host/nodes/Paragraph';
import { capitalize } from '../utils';
import {
  SingleOrMulti,
  TextTransform,
  OnLayoutHook,
  MouseEventHandler,
} from '../types';

export type Props = {
  children: SingleOrMulti<number | string | JSX.Element>;
  color?: string;
  bgColor?: string;
  textTransform?: TextTransform;
  dim?: boolean;
  underline?: boolean;
  lineThrough?: boolean;
  bold?: boolean;
  italic?: boolean;
  onLayout?: OnLayoutHook;
  onClick?: MouseEventHandler;
  onWheel?: MouseEventHandler;
};

export default function View(props: Props) {
  return (
    // @ts-ignore
    <Paragraph.TAG
      onLayout={props.onLayout}
      onClick={props.onClick}
      onWheel={props.onWheel}
      style={{
        color: props.color,
        bgColor: props.bgColor,
        modifiers: reduceModifiers(props),
      }}
    >
      {React.Children.map(props.children, child => {
        if (React.isValidElement(child)) {
          return React.cloneElement(child, {
            textTransform: props.textTransform,
          } as any);
        }

        switch (props.textTransform) {
          case 'capitalize': {
            return capitalize(child.toString());
          }
          case 'lowercase': {
            return child.toString().toLowerCase();
          }
          case 'uppercase': {
            return child.toString().toUpperCase();
          }
          case 'none':
          default:
            return child;
        }
      })}
    </Paragraph.TAG>
  );
}

function reduceModifiers(props: Props) {
  const modifiers = [];
  if (props.bold) {
    modifiers.push('bold');
  }
  if (props.italic) {
    modifiers.push('italic');
  }
  if (props.dim) {
    modifiers.push('dim');
  }
  if (props.underline) {
    modifiers.push('underline');
  }
  if (props.lineThrough) {
    modifiers.push('strikethrough');
  }
  return modifiers.length ? modifiers : undefined;
}
