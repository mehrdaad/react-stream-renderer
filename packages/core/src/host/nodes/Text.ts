import assert from 'assert';
import stripAnsi from 'strip-ansi';
import Base from './Base';

export type TextStyle = {
  color?: string;
  bgColor?: string;
  modifiers?: string[];
};

export default class Text extends Base<never> {
  static TAG = 'TEXT_NODE';

  private body: string = '';

  setBody(body: string) {
    assert(!body.includes('\n'), 'Text body cannot have new line characters');
    this.body = body;
    this.layoutNode.setStyle({
      width: stripAnsi(this.body).length,
      height: 1,
    });
  }

  getBody(): string {
    return this.body;
  }
}
