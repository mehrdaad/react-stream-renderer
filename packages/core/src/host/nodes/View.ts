import assert from 'assert';
import Base from './Base';
import Text from './Text';
import { Style as LayoutStyle, PositionType } from '../../layout';
import { Layout } from '../../layout';
import { MouseEvent } from '../../types';

export default class View extends Base<View | Text> {
  static TAG = 'VIEW_NODE';

  setLayoutStyle(style: LayoutStyle & { zIndex?: number }) {
    const isAbsolute = style.positionType === PositionType.Absolute;
    if (this.isAbsolute !== isAbsolute) {
      this.isAbsolute = isAbsolute;
      if (style.zIndex) {
        this.zIndex = style.zIndex;
      } else {
        this.zIndex = isAbsolute ? 1 : 0;
      }
      for (let i = 0; i < this.children.length; i++) {
        this.children[i].isAbsolute = this.isAbsolute;
        this.children[i].zIndex = this.zIndex;
      }
    }

    this.layoutNode.setStyle({
      ...this.layoutNode.getStyle(),
      ...style,
    });
  }

  findChild(child: View | Text) {
    return this.children.indexOf(child);
  }

  private updateLayoutNode() {
    // Remove all previous children layout nodes.
    // After each removal `childCount` will return a new value - if greater than 0,
    // the next child to remove will be moved to the top, at index 0.
    while (this.layoutNode.childCount) {
      this.layoutNode.removeChildAtIndex(0);
    }
    for (let i = 0; i < this.children.length; i++) {
      this.children[i].isAbsolute = this.isAbsolute;
      this.children[i].zIndex = this.zIndex;
      this.layoutNode.addChild(this.children[i].layoutNode);
    }
  }

  prependChild(child: View | Text, position?: number) {
    child.parent = this;
    if (position !== undefined) {
      assert(
        position >= 0 && position <= this.children.length - 1,
        `child position out of bounds: ${position}`
      );
      this.children.splice(position, 0, child);
    } else {
      this.children.unshift(child);
    }

    this.updateLayoutNode();
  }

  appendChild(child: View | Text, position?: number) {
    child.parent = this;
    if (position !== undefined) {
      assert(
        position >= 0 && position <= this.children.length,
        `child position out of bounds: ${position}`
      );

      this.children.splice(position + 1, 0, child);
    } else {
      this.children.push(child);
    }

    this.updateLayoutNode();
  }

  removeChild(child: View | Text) {
    const nodeIndex = this.children.indexOf(child);
    assert(nodeIndex >= 0, 'child not found');
    delete child.parent;
    delete child.layoutNode;
    this.children.splice(nodeIndex, 1);
    this.layoutNode.removeChildAtIndex(nodeIndex);
  }

  notifyOnLayoutHook(
    layout: Layout,
    { offsetX, offsetY }: { offsetX: number; offsetY: number }
  ) {
    super.notifyOnLayoutHook(layout, { offsetX, offsetY });
    for (let i = 0; i < this.children.length; i++) {
      this.children[i].notifyOnLayoutHook(layout.child(i), {
        offsetX: offsetX + layout.x,
        offsetY: offsetY + layout.y,
      });
    }
  }

  notifyOnClickHook(event: MouseEvent) {
    const handled = this.children.some(child => child.notifyOnClickHook(event));
    if (!handled) {
      return Boolean(super.notifyOnClickHook(event));
    }
    return false;
  }
}
