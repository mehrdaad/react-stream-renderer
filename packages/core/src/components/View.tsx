import React from 'react';
import { default as HostView } from '../host/nodes/View';
import {
  Style as LayoutStyle,
  Display,
  PositionType,
  FlexDirection,
  FlexWrap,
  Overflow,
  AlignItems,
  AlignContent,
  AlignSelf,
  JustifyContent,
} from '../layout';
import { SingleOrMulti, OnLayoutHook, MouseEventHandler } from '../types';

export type Props = {
  children?: SingleOrMulti<0 | false | null | undefined | JSX.Element>;
  bgColor?: string;
  /** Layout props */
  display?: 'flex' | 'none';
  position?: 'relative' | 'absolute';
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse';
  flexWrap?: 'no-wrap' | 'wrap' | 'wrap-reverse';
  flexGrow?: number;
  flexShrink?: number;
  flexBasis?: number | string;
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch';
  alignSelf?:
    | 'auto'
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'baseline'
    | 'stretch';
  alignContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'stretch'
    | 'space-between'
    | 'space-around';
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  top?: number | string;
  right?: number | string;
  bottom?: number | string;
  left?: number | string;
  zIndex?: number;
  marginTop?: number | string;
  marginRight?: number | string;
  marginBottom?: number | string;
  marginLeft?: number | string;
  paddingTop?: number | string;
  paddingRight?: number | string;
  paddingBottom?: number | string;
  paddingLeft?: number | string;
  width?: number | string;
  height?: number | string;
  minWidth?: number | string;
  minHeight?: number | string;
  maxWidth?: number | string;
  maxHeight?: number | string;
  /** Hooks & handlers */
  onLayout?: OnLayoutHook;
  onClick?: MouseEventHandler;
  onWheel?: MouseEventHandler;
};

export default function View(props: Props) {
  const { children, onLayout, bgColor } = props;
  return (
    // @ts-ignore
    <HostView.TAG
      onLayout={onLayout}
      onClick={props.onClick}
      onWheel={props.onWheel}
      layoutStyle={reduceLayoutStyle(props)}
      style={{ bgColor }}
    >
      {children}
    </HostView.TAG>
  );
}

export function reduceLayoutStyle(
  props: Props
): LayoutStyle & { zIndex: number } {
  return {
    display: props.display === 'none' ? Display.None : Display.Flex,
    positionType:
      props.position === 'absolute'
        ? PositionType.Absolute
        : PositionType.Relative,
    flexDirection: {
      row: FlexDirection.Row,
      'row-reverse': FlexDirection.RowReverse,
      column: FlexDirection.Column,
      'column-reverse': FlexDirection.ColumnReverse,
    }[props.flexDirection || 'row'],
    flexWrap:
      props.flexWrap &&
      {
        'no-wrap': FlexWrap.NoWrap,
        wrap: FlexWrap.Wrap,
        'wrap-reverse': FlexWrap.WrapReverse,
      }[props.flexWrap],
    overflow: Overflow.Hidden,
    alignItems:
      props.alignItems &&
      {
        'flex-start': AlignItems.FlexStart,
        'flex-end': AlignItems.FlexEnd,
        center: AlignItems.Center,
        baseline: AlignItems.Baseline,
        stretch: AlignItems.Stretch,
      }[props.alignItems],
    alignSelf:
      props.alignSelf &&
      {
        auto: AlignSelf.Auto,
        'flex-start': AlignSelf.FlexStart,
        'flex-end': AlignSelf.FlexEnd,
        center: AlignSelf.Center,
        baseline: AlignSelf.Baseline,
        stretch: AlignSelf.Stretch,
      }[props.alignSelf],
    alignContent:
      props.alignContent &&
      {
        'flex-start': AlignContent.FlexStart,
        'flex-end': AlignContent.FlexEnd,
        center: AlignContent.Center,
        stretch: AlignContent.Stretch,
        'space-between': AlignContent.SpaceBetween,
        'space-around': AlignContent.SpaceAround,
      }[props.alignContent],
    justifyContent:
      props.justifyContent &&
      {
        'flex-start': JustifyContent.FlexStart,
        'flex-end': JustifyContent.FlexEnd,
        center: JustifyContent.Center,
        'space-between': JustifyContent.SpaceBetween,
        'space-around': JustifyContent.SpaceAround,
        'space-evenly': JustifyContent.SpaceEvenly,
      }[props.justifyContent],
    start: props.left,
    end: props.right,
    top: props.top,
    bottom: props.bottom,
    marginStart: props.marginLeft,
    marginEnd: props.marginRight,
    marginTop: props.marginTop,
    marginBottom: props.marginBottom,
    paddingStart: props.paddingLeft,
    paddingEnd: props.paddingRight,
    paddingTop: props.paddingTop,
    paddingBottom: props.paddingBottom,
    flexGrow: props.flexGrow,
    flexShrink: props.flexShrink,
    flexBasis: props.flexBasis,
    width: props.width,
    height: props.height,
    minWidth: props.minWidth,
    minHeight: props.minHeight,
    maxWidth: props.maxWidth,
    maxHeight: props.maxHeight,
    zIndex: props.position === 'absolute' ? props.zIndex || 1 : 0,
  };
}
