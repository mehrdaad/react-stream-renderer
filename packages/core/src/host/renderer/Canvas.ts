import Paragraph from '../nodes/Paragraph';
import View from '../nodes/View';
import Text from '../nodes/Text';
import { Layout } from '../../layout';
import { Style } from '../../types';

export type Pixel = {
  char: string;
  style?: Style;
  z: number;
};

export default class Canvas {
  pixels: Pixel[][] = [];
  width: number = 0;
  height: number = 0;
  x: number = 0;
  y: number = 0;

  private resize(layout: Layout) {
    for (let y = 0; y < layout.height; y += 1) {
      this.pixels[y] = [];
      for (let x = 0; x < layout.width; x += 1) {
        this.pixels[y][x] = {
          char: '',
          z: -Number.MAX_VALUE,
        };
      }
    }

    this.width = layout.width;
    this.height = layout.height;
    this.x = layout.x;
    this.y = layout.y;
  }

  private fillPixel(
    { x, y, z }: { x: number; y: number; z: number },
    char: string | undefined,
    style?: Style
  ) {
    // Pixel is below already stored pixel
    if (z < this.pixels[y][x].z) {
      return;
    }

    if (char) {
      this.pixels[y][x].char = char;
    }

    if (this.pixels[y][x].z === z) {
      if (style) {
        this.pixels[y][x].style = {
          color:
            style.color ||
            (this.pixels[y][x].style && this.pixels[y][x].style!.color),
          bgColor:
            style.bgColor ||
            (this.pixels[y][x].style && this.pixels[y][x].style!.bgColor),
          modifiers:
            style.modifiers ||
            (this.pixels[y][x].style && this.pixels[y][x].style!.modifiers),
        };
      }
    } else {
      this.pixels[y][x].style = {
        color: style ? style.color : undefined,
        bgColor: style ? style.bgColor : undefined,
        modifiers: style ? style.modifiers : undefined,
      };
    }

    this.pixels[y][x].z = z;
  }

  fill(
    node: View | Text | Paragraph,
    layout: Layout,
    { parentZ }: { parentZ: number }
  ) {
    this.resize(layout);

    const body = node instanceof Text && node.getBody();

    for (let y = 0; y < layout.height; y++) {
      // Body X index
      let bx = 0;
      for (let x = 0; x < layout.width; x += 1) {
        this.fillPixel(
          { x, y, z: node.isAbsolute ? node.zIndex : parentZ },
          body ? body[bx] : undefined,
          node.style
        );
        bx += 1;
      }
    }
  }

  mergeChildCanvas(childCanvas: Canvas) {
    for (
      let cy = childCanvas.y, y = 0;
      cy < childCanvas.y + childCanvas.height && cy < this.height;
      cy += 1, y += 1
    ) {
      if (cy < 0) {
        continue;
      }

      for (
        let cx = childCanvas.x, x = 0;
        cx < childCanvas.x + childCanvas.width && cx < this.width;
        cx += 1, x += 1
      ) {
        if (cx < 0) {
          continue;
        }

        const pixel = childCanvas.pixels[y][x];
        this.fillPixel({ x: cx, y: cy, z: pixel.z }, pixel.char, pixel.style);
      }
    }
  }
}
