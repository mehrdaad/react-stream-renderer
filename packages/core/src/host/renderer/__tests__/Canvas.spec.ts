import Canvas from '../Canvas';
import View from '../../nodes/View';
import Text from '../../nodes/Text';
import { Layout } from '../../../layout';

describe('Canvas', () => {
  it('should fill pixels', () => {
    let canvas = new Canvas();
    let layout: Layout;

    const emptyView = new View();
    emptyView.setLayoutStyle({ width: 0, height: 0 });
    layout = emptyView.layoutNode.computeLayout({
      width: null,
      height: null,
    });

    canvas.resize(layout);
    canvas.fill(emptyView, layout, { z: 0 });
    expect(canvas.width).toBe(0);
    expect(canvas.height).toBe(0);
    expect(canvas.pixels).toEqual([]);

    canvas = new Canvas();
    const viewStyleless = new View();
    viewStyleless.setLayoutStyle({ width: 5, height: 2 });
    layout = viewStyleless.layoutNode.computeLayout({
      width: null,
      height: null,
    });

    canvas.resize(layout);
    canvas.fill(viewStyleless, layout, { z: 0 });
    expect(canvas.width).toBe(5);
    expect(canvas.height).toBe(2);
    const rowStyleless = new Array(5).fill(null).map(() => ({
      char: '',
      z: 0,
      style: {
        bgColor: undefined,
        color: undefined,
        modifiers: undefined,
      },
    }));
    expect(canvas.pixels).toEqual([rowStyleless, rowStyleless]);

    canvas = new Canvas();
    const viewStyled = new View();
    viewStyled.setLayoutStyle({ width: 5, height: 2 });
    viewStyled.style = { bgColor: 'red' };
    layout = viewStyled.layoutNode.computeLayout({
      width: null,
      height: null,
    });

    canvas.resize(layout);
    canvas.fill(viewStyled, layout, { z: 0 });
    expect(canvas.width).toBe(5);
    expect(canvas.height).toBe(2);
    const rowStyled = new Array(5).fill(null).map(() => ({
      char: '',
      z: 0,
      style: {
        color: undefined,
        bgColor: 'red',
        modifiers: undefined,
      },
    }));
    expect(canvas.pixels).toEqual([rowStyled, rowStyled]);
  });

  it('should fill pixels with body', () => {
    const canvas = new Canvas();
    const text = new Text();
    text.setBody('Hello World');
    text.style = { color: 'green' };
    const layout = text.layoutNode.computeLayout({
      width: null,
      height: null,
    });

    canvas.resize(layout);
    canvas.fill(text, layout, { z: 0 });
    expect(canvas.width).toBe(11);
    expect(canvas.height).toBe(1);
    expect(canvas.pixels).toEqual([
      new Array(11).fill(null).map((_, i) => ({
        char: 'Hello World'[i],
        z: 0,
        style: {
          color: 'green',
          bgColor: undefined,
          modifiers: undefined,
        },
      })),
    ]);
  });

  it('should fill pixels and trim on top-edge', () => {
    const canvas = new Canvas();
    const view = new View();
    const child = new View();
    view.setLayoutStyle({ paddingTop: -1 });
    child.setLayoutStyle({ width: 1, height: 2 });
    view.appendChild(child);
    const layout = view.layoutNode.computeLayout({
      width: null,
      height: null,
    });

    canvas.resize(layout.child(0));
    canvas.fill(child, layout.child(0), { z: 0 });
    expect(canvas.width).toBe(1);
    expect(canvas.height).toBe(1);
    expect(canvas.pixels).toEqual([
      [
        {
          char: '',
          z: 0,
          style: {
            color: undefined,
            bgColor: undefined,
            modifiers: undefined,
          },
        },
      ],
    ]);
  });

  it('should fill pixels and trim on left-edge', () => {
    const canvas = new Canvas();
    const view = new View();
    const child = new View();
    view.setLayoutStyle({ paddingStart: -1 });
    child.setLayoutStyle({ width: 2, height: 1 });
    view.appendChild(child);
    const layout = view.layoutNode.computeLayout({
      width: null,
      height: null,
    });

    canvas.resize(layout.child(0));
    canvas.fill(child, layout.child(0), { z: 0 });
    expect(canvas.width).toBe(1);
    expect(canvas.height).toBe(1);
    expect(canvas.pixels).toEqual([
      [
        {
          char: '',
          z: 0,
          style: {
            color: undefined,
            bgColor: undefined,
            modifiers: undefined,
          },
        },
      ],
    ]);
  });
});
