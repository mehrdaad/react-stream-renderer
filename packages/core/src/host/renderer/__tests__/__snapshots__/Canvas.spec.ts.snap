// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas should fill pixels and trim on bottom-edge 1`] = `
Array [
  Array [
    Object {
      "char": "H",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "e",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "o",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
]
`;

exports[`Canvas should fill pixels and trim on left-edge 1`] = `
Array [
  Array [
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
  Array [
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "o",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
  Array [
    Object {
      "char": "r",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "d",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
  Array [
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
]
`;

exports[`Canvas should fill pixels and trim on right-edge 1`] = `
Array [
  Array [
    Object {
      "char": "H",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "e",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "green",
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
  Array [
    Object {
      "char": "W",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "o",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "r",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
]
`;

exports[`Canvas should fill pixels and trim on top-edge 1`] = `
Array [
  Array [
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "W",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "o",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "r",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "d",
      "style": Object {
        "bgColor": "gray",
        "color": "red",
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
  Array [
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "",
      "style": Object {
        "bgColor": "gray",
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
]
`;

exports[`Canvas should merge canvases with absolute elements 1`] = `
Array [
  Array [
    Object {
      "char": "A",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "b",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "s",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "o",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "u",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "t",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "e",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 1,
    },
    Object {
      "char": "r",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "l",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
    Object {
      "char": "d",
      "style": Object {
        "bgColor": undefined,
        "color": undefined,
        "modifiers": undefined,
      },
      "z": 0,
    },
  ],
]
`;
